### PR TITLE
Add date functionality to Expense

### PR DESCRIPTION
**Pull request by Progresst-8**
issue #57 

Add date attribute to expense
Add DateUtils class for date utility
Update and add JUnit tests for Expense and ExpenseList
Update add, edit, list, load, save commands
Update and add validation and exception handling to various methods in Expense and ExpenseList

### CODE REVIEW

File: src/main/java/wheresmymoney/DateUtils.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    
+    public static boolean isInDateFormat(String dateAsString) {
+        try {
+            LocalDate.parse(dateAsString, formatter);
+            return true;
+        } catch (DateTimeParseException e) {
+            return false;
+        }
+    }
+    public static LocalDate getLocalDateNow() {
```
Comment (KHsienKit) [inline src/main/java/wheresmymoney/DateUtils.java]: change function name to getCurrentDate

File: src/main/java/wheresmymoney/DateUtils.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        return LocalDate.now();
+    }
+    public static LocalDate stringToDateFormat(String dateAsString) throws WheresMyMoneyException {
```
Comment (KHsienKit) [inline src/main/java/wheresmymoney/DateUtils.java:24]: Change function name to stringToDate

File: src/main/java/wheresmymoney/ExpenseList.java [hunk 3]
```diff
@@ -32,26 +32,35 @@ public boolean isEmpty() {
         return expenses.isEmpty();
     }
 
-    public Expense getExpenseAtIndex(int i) {
-        return expenses.get(i);
+    public Expense getExpenseAtIndex(int i) throws WheresMyMoneyException {
```
Comment (KHsienKit) [inline src/main/java/wheresmymoney/ExpenseList.java:35]: instead of using "i" in your params, use the full word "index"

**Review by KHsienKit — CHANGES_REQUESTED** [review]

File: src/main/java/wheresmymoney/command/AddCommand.java [hunk 4]
```diff
@@ -19,8 +19,13 @@ public void execute(ExpenseList expenseList) throws WheresMyMoneyException {
             float price = Float.parseFloat(argumentsMap.get(Parser.ARGUMENT_PRICE));
             String description = argumentsMap.get(Parser.ARGUMENT_DESCRIPTION);
             String category = argumentsMap.get(Parser.ARGUMENT_CATEGORY);
-            expenseList.addExpense(price, description, category);
-        }catch(NullPointerException | NumberFormatException e){
+            if (argumentsMap.containsKey(Parser.ARGUMENT_CATEGORY)) {
```
Comment (Hackin7) [inline src/main/java/wheresmymoney/command/AddCommand.java:22]: Hmm I don't really get this line

Comment (Hackin7) [reply src/main/java/wheresmymoney/command/AddCommand.java:22]: does adding a date need a category?

File: src/main/java/wheresmymoney/command/ListCommand.java [hunk 5]
```diff
@@ -24,7 +24,8 @@ private ArrayList<Expense> getExpensesToDisplay(ExpenseList expenseList) {
         }
     }
 
-    private void displayExpenses(ArrayList<Expense> expensesToDisplay, ExpenseList expenseList) {
+    private void displayExpenses(ArrayList<Expense> expensesToDisplay, ExpenseList expenseList)
```
Comment (Hackin7) [inline src/main/java/wheresmymoney/command/ListCommand.java:27]: this function is not updated to display expenses with date - we should add that in

File: src/main/java/wheresmymoney/command/AddCommand.java (hunk 4 above)
Comment (Progresst-8) [reply src/main/java/wheresmymoney/command/AddCommand.java]: corrected

File: src/main/java/wheresmymoney/command/ListCommand.java (hunk 5 above)
Comment (Progresst-8) [reply src/main/java/wheresmymoney/command/ListCommand.java:27]: corrected

**Review by Hackin7 — APPROVED** [review]
Looks good to me, if there are any changes we want to make to the listing (considering the other user stories) we would just make a new PR for it

File: src/main/java/wheresmymoney/ExpenseList.java (hunk 3 above)
Comment (Hackin7) [reply src/main/java/wheresmymoney/ExpenseList.java]: changed

**Review by KHsienKit — APPROVED** [review]
Looks good

**Pull request merged**